File: Core/Workspace/CSharp/Adapters/Allors.Workspace.Adapters.Remote/Workspace/WorkspaceRecord.cs
// <copyright file="RemoteWorkspaceObject.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>

namespace Allors.Workspace.Adapters.Remote
{
    using System.Collections.Generic;
    using System.Linq;
    using Meta;

    internal class WorkspaceRecord : IRecord
    {
        private readonly Database database;
        private readonly IClass @class;
        private readonly long id;
        private readonly IReadOnlyDictionary<IRelationType, object> roleByRelationType;

        internal WorkspaceRecord(Database database, long id, IClass @class, long version, IReadOnlyDictionary<IRelationType, object> roleByRelationType)
        {
            this.database = database;
            this.id = id;
            this.@class = @class;
            this.Version = version;
            this.roleByRelationType = this.Import(roleByRelationType).ToDictionary(v => v.Key, v => v.Value);
        }

        public WorkspaceRecord(WorkspaceRecord originalWorkspaceRecord, IReadOnlyDictionary<IRelationType, object> changedRoleByRoleType)
        {
            this.database = originalWorkspaceRecord.database;
            this.id = originalWorkspaceRecord.id;
            this.@class = originalWorkspaceRecord.@class;
            this.Version = ++originalWorkspaceRecord.Version;

            this.roleByRelationType = this.Import(changedRoleByRoleType, originalWorkspaceRecord.roleByRelationType).ToDictionary(v => v.Key, v => v.Value);
        }

        public long Version { get; private set; }

        public object GetRole(IRoleType roleType)
        {
            object @object = null;
            _ = (this.roleByRelationType?.TryGetValue(roleType.RelationType, out @object));
            return @object;
        }

        private IEnumerable<KeyValuePair<IRelationType, object>> Import(IReadOnlyDictionary<IRelationType, object> changedRoleByRoleType, IReadOnlyDictionary<IRelationType, object> originalRoleByRoleType = null)
        {
            foreach (var roleType in this.@class.WorkspaceOriginRoleTypes)
            {
                var relationType = roleType.RelationType;

                if (changedRoleByRoleType.TryGetValue(relationType, out var role))
                {
                    if (role != null)
                    {
                        if (roleType.ObjectType.IsUnit)
                        {
                            yield return new KeyValuePair<IRelationType, object>(relationType, role);
                        }
                        else
                        {
                            if (roleType.IsOne)
                            {
                                yield return new KeyValuePair<IRelationType, object>(relationType, ((Strategy)role).Id);

                            }
                            else
                            {
                                var roles = (Strategy[])role;
                                if (roles.Length > 0)
                                {
                                    yield return new KeyValuePair<IRelationType, object>(relationType, roles.Select(v => v.Id).ToArray());
                                }
                            }
                        }
                    }
                }
                else if (originalRoleByRoleType?.TryGetValue(roleType.RelationType, out role) == true)
                {
                    yield return new KeyValuePair<IRelationType, object>(relationType, role);
                }
            }
        }
    }
}

File: typescript/modules/libs/system/workspace/adapters/src/lib/workspace/workspace-record.ts
import { Class, RelationType, RoleType } from '@allors/workspace/meta/system';
import { IRecord } from '../irecord';
import { Strategy } from '../session/strategy';

export class WorkspaceRecord implements IRecord {
  constructor(public cls: Class, public id: number, public version: number, public roleByRelationType: Map<RelationType, any>) {}

  static fromOriginal(originalRecord: WorkspaceRecord, changedRoleByRoleType: Map<RelationType, unknown>): WorkspaceRecord {
    const cls = originalRecord.cls;

    const originalRoleByRelationType = originalRecord.roleByRelationType;

    const roleByRelationType = new Map(
      Array.from(cls.roleTypes)
        .filter((roleType) => changedRoleByRoleType.has(roleType.relationType) || originalRoleByRelationType.has(roleType.relationType))
        .map((roleType) => {
          const relationType = roleType.relationType;
          if (changedRoleByRoleType.has(relationType)) {
            const role = changedRoleByRoleType.get(relationType);
            if (roleType.objectType.isUnit) {
              return [relationType, role];
            } else {
              if (roleType.isOne) {
                return [relationType, (role as Strategy).id];
              } else {
                const roles = role as Strategy[];
                return [relationType, roles.map((v) => v.id)];
              }
            }
          } else {
            const role = originalRecord.roleByRelationType.get(relationType) as unknown;
            return [relationType, role];
          }
        })
    );

    return new WorkspaceRecord(cls, originalRecord.id, ++originalRecord.version, roleByRelationType);
  }

  getRole(roleType: RoleType): unknown {
    return this.roleByRelationType?.get(roleType.relationType);
  }
}
